src/Server/Completion/TokenContainer.php

```diff
@@ -0,0 +1,80 @@
+<?php
+namespace LanguageServer\Server\Completion;
+
+use LanguageServer\Protocol\Position;
+use LanguageServer\Protocol\Range;
+
+class TokenContainer
+{
+
+    const T_SINGLE_CHAR = 10000;
+
+    /**
+     *
+     * @var \LanguageServer\Server\Completion\Token[]
+     */
+    private $tokens;
+
+    public function __construct(array $tokens)
```

Comment:
Another use case is typing just:

```php
<?php
abstract
```

In this case the PHP parser returns an empty AST tree.

The only way to find all situations when AST tree is not enough seems to be by trial and error. So, it looks reasonable to rely on tokens than on AST.